clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -576,7 +576,7 @@ private void completeBatch(ProducerBatch batch, ProduceResponse.PartitionRespons
                                long now, long throttleUntilTimeMs) {
         Errors error = response.error;
 
-        if (error == Errors.MESSAGE_TOO_LARGE && batch.recordCount > 1 &&
+        if (error == Errors.MESSAGE_TOO_LARGE && batch.recordCount > 1 && batch.finalState() == null &&
```

Comment:
Can we please add a method `isDone` which is basically `finalState() != null)` and then use it here and everywhere else we do `batch.finalState() == null`? It would be much easier to understand.